src/sentry/static/sentry/app/components/smartSearchBar.jsx

```diff
@@ -0,0 +1,497 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import Reflux from 'reflux';
+import _ from 'lodash';
+import classNames from 'classnames';
+import createReactClass from 'create-react-class';
+import styled from 'react-emotion';
+
+import {t} from 'app/locale';
+import MemberListStore from 'app/stores/memberListStore';
+import SearchDropdown from 'app/views/stream/searchDropdown';
+import withOrganization from 'app/utils/withOrganization';
+
+export function addSpace(query = '') {
+  if (query.length !== 0 && query[query.length - 1] !== ' ') {
+    return query + ' ';
+  } else {
+    return query;
+  }
+}
+
+export function removeSpace(query = '') {
+  if (query[query.length - 1] === ' ') {
+    return query.slice(0, query.length - 1);
+  } else {
+    return query;
+  }
+}
+class SmartSearchBar extends React.Component {
+  static propTypes = {
+    defaultQuery: PropTypes.string,
+    query: PropTypes.string,
+    defaultSearchItems: PropTypes.array.isRequired,
+    disabled: PropTypes.bool,
+    placeholder: PropTypes.string,
+
+    // Map of tags
+    supportedTags: PropTypes.object,
+
+    onGetTagValues: PropTypes.func,
+
+    onSearch: PropTypes.func,
+    // If true, excludes the environment tag from the autocompletion list
+    // This is because we don't want to treat environment as a tag in some places
+    // such as the stream view where it is a top level concept
+    excludeEnvironment: PropTypes.bool,
+  };
+
+  /**
+     * Given a query, and the current cursor position, return the string-delimiting
+     * index of the search term designated by the cursor.
+     */
+  static getLastTermIndex = (query, cursor) => {
+    // TODO: work with quoted-terms
+    let cursorOffset = query.slice(cursor).search(/\s|$/);
+    return cursor + (cursorOffset === -1 ? 0 : cursorOffset);
+  };
+
+  /**
+     * Returns an array of query terms, including incomplete terms
+     *
+     * e.g. ["is:unassigned", "browser:\"Chrome 33.0\"", "assigned"]
+     */
+  static getQueryTerms = (query, cursor) => {
+    return query.slice(0, cursor).match(/\S+:"[^"]*"?|\S+/g);
+  };
```

Comment:
Is there a particular reason these are static class methods?

Reply:
This was from previous code